sycl/include/CL/sycl/detail/aligned_allocator.hpp

```diff
@@ -0,0 +1,71 @@
+//==------------ aligned_allocator.hpp - SYCL standard header file ---------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/cl.h>
+#include <CL/sycl/detail/cnri.h>
+#include <CL/sycl/range.hpp>
+
+#include <cstring>
+#include <memory>
+#include <vector>
+
+namespace cl {
+namespace sycl {
+template <typename T, size_t Alignment>
+class aligned_allocator {
+public:
+  using value_type = T;
+  using pointer = T*;
+  using const_pointer = const T*;
+  using reference = T&;
+  using const_reference = const T&;
+
+public:
+  template<typename U>
+  struct rebind {
+    typedef aligned_allocator<U, Alignment> other;
+  };
+
+  aligned_allocator() = default;
+  aligned_allocator(const aligned_allocator&) = default;
+
+  void construct(pointer Ptr, const_reference Val) {
+    new (Ptr) value_type(Val);
+  }
+
+  void destroy(pointer Ptr) { Ptr->~value_type(); }
+
+  pointer address(reference Val) const { return &Val; }
+  const_pointer address(const_reference Val) { return &Val; }
+
+  pointer allocate(size_t Size) {
+    Size += Alignment - Size % Alignment;
+    pointer Result = reinterpret_cast<pointer>(
+      aligned_alloc(Alignment, Size * sizeof(value_type)));
+    if (!Result)
+      throw std::bad_alloc();
+    return Result;
+  }
+
+  void deallocate(pointer Ptr, size_t size) {
+    if (Ptr)
+      free(Ptr);
+  }
+
+  aligned_allocator<T, Alignment>& operator =
+    (const aligned_allocator& other) { return *this; }
```

Comment:
Why do you need a copy assignment operator?

Reply:
I don't really need it in this patch. But I'd prefer to follow spec for allocators to use it if someone wants to create a custom container.

Reply:
But I have the feeling that the compiler will synthesize this copy assignment operator for you without having to do anything. Is there something different from the default assignment?

Reply:
Done